### PR TITLE
[AArch64][SelectionDAG] Vector splitting and promotion for histogram intrinsic

### PR DESCRIPTION
**Pull request by DevM-uk**
Adds support for wider-than-legal vector types for the histogram intrinsic (llvm.experimental.vector.histogram.add) by splitting the vector.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -25434,6 +25436,41 @@ performScalarToVectorCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
   return NVCAST;
 }
 
+static SDValue performHistogramCombine(SDNode *N,
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:25439]: Rather than doing this as a target-specific DAG combine, this should be done as generic type legalisation in LegalizeVectorTypes.cpp. See `DAGTypeLegalizer::SplitVectorOperand` for how this is done for other operations.

Comment (DevM-uk) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: You're right. I've moved the splitting to `DAGTypeLegalizer::SplitVecOp_VECTOR_HISTOGRAM` and added/moved the `INC` operand promotion to `DAGTypeLegalizer::PromoteIntOp_VECTOR_HISTOGRAM`.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 2]
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
                        DAG.getNode(ISD::ADD, DL, ResVT, VLo, ResHi));
 }
 
+SDValue DAGTypeLegalizer::SplitVecOp_VECTOR_HISTOGRAM(SDNode *N) {
+  MaskedHistogramSDNode *HG = cast<MaskedHistogramSDNode>(N);
+  SDLoc DL(HG);
+  SDValue Chain = HG->getChain();
```
Comment (sdesmalen-arm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:4283]: nit: `Chain` has one use, just inline.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 3]
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
...
+  SDValue Inc = HG->getInc();
+  SDValue Mask = HG->getMask();
+  SDValue Ptr = HG->getBasePtr();
+  SDValue Index = HG->getIndex();
```
Comment (sdesmalen-arm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:4287]: nit: `Index` has one use, just inline.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 4]
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
                        DAG.getNode(ISD::ADD, DL, ResVT, VLo, ResHi));
 }
 
+SDValue DAGTypeLegalizer::SplitVecOp_VECTOR_HISTOGRAM(SDNode *N) {
+  MaskedHistogramSDNode *HG = cast<MaskedHistogramSDNode>(N);
+  SDLoc DL(HG);
+  SDValue Chain = HG->getChain();
+  SDValue Inc = HG->getInc();
+  SDValue Mask = HG->getMask();
```
Comment (sdesmalen-arm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:4285]: nit: `Mask` has one use, just inline.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 5]
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
...
+  SDValue Scale = HG->getScale();
+  SDValue IntID = HG->getIntID();
+  EVT MemVT = HG->getMemoryVT();
+  MachineMemOperand *MMO = HG->getMemOperand();
+  ISD::MemIndexType IndexType = HG->getIndexType();
+
+  SDValue IndexLo, IndexHi, MaskLo, MaskHi;
+  std::tie(IndexLo, IndexHi) = DAG.SplitVector(Index, DL);
+  std::tie(MaskLo, MaskHi) = DAG.SplitVector(Mask, DL);
+  SDValue HistogramOpsLo[] = {Chain, Inc, MaskLo, Ptr, IndexLo, Scale, IntID};
```
Comment (sdesmalen-arm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:4297]: nit:
```suggestion
  SDValue OpsLo[] = {Chain, Inc, MaskLo, Ptr, IndexLo, Scale, IntID};
```
or just inline the initializer-sequence into the `getMaskedHistogram` directly.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 6]
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
...
+  SDValue HChain = DAG.getMaskedHistogram(DAG.getVTList(MVT::Other), MemVT, DL,
```
Comment (sdesmalen-arm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:4298]: nit:
```suggestion
  SDValue Lo = DAG.getMaskedHistogram(DAG.getVTList(MVT::Other), MemVT, DL,
```

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 7]
```diff
@@ -1777,10 +1777,12 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
 
     // Histcnt is SVE2 only
     if (Subtarget->hasSVE2()) {
-      setOperationAction(ISD::EXPERIMENTAL_VECTOR_HISTOGRAM, MVT::Other,
+      setOperationAction(ISD::EXPERIMENTAL_VECTOR_HISTOGRAM, MVT::i8, Promote);
+      setOperationAction(ISD::EXPERIMENTAL_VECTOR_HISTOGRAM, MVT::i16, Promote);
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: These two `setOperationActions`s are unnecessary, because i8 and i16 are illegal types, so the legalizer will already try to legalize the operation, regardless of this `setOperationAction`.

**Review by sdesmalen-arm — APPROVED** [review]
Left a few nits, but otherwise looks good to me!

**Comment by sdesmalen-arm** [comment]
On the commit message, the patch is also doing type promotion, so maybe you want to rephrase it to something slightly more generic, e.g. `Vector splitting and promotion for histogram` ?

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp (hunk 6 above)
Comment (DevM-uk) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp]: Done.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp (hunk 5 above)
Comment (DevM-uk) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp]: I've made the suggested change. Operand lists are passed in using arrays in other places (such as with `getMaskedScatter`s) rather than passing the initializer-sequence inline.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp (hunk 4 above)
Comment (DevM-uk) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp]: Done.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp (hunk 3 above)
Comment (DevM-uk) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp]: Done.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp (hunk 2 above)
Comment (DevM-uk) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp]: Done.
